src/test/management/list.spec.ts

```diff
@@ -119,4 +131,204 @@ describe("list", () => {
+      const appCounts = 10;
+      const pageSize = 5;
+      const nextPageToken = "next-page-token";
+      const expectedAppList = generateAppList(appCounts, AppPlatform.PLATFORM_UNSPECIFIED);
+      const expectedError = new Error("HTTP Error 400: unexpected error");
+      const firstCallStub = listAppsApiRequestStub(undefined, pageSize).resolves({
+        body: { apps: expectedAppList.slice(0, pageSize), nextPageToken },
+      });
+      const secondCallStub = listAppsApiRequestStub(undefined, pageSize, nextPageToken).rejects(
+        expectedError
+      );
+
+      let err;
+      try {
+        await listFirebaseApps(PROJECT_ID, undefined, pageSize);
+      } catch (e) {
+        err = e;
+      }
+      expect(err.message).to.equal(
+        "Failed to list Firebase apps. See firebase-debug.log for more info."
+      );
+      expect(err.original).to.equal(expectedError);
+      expect(firstCallStub).to.be.calledOnce;
+      expect(secondCallStub).to.be.calledOnce;
+    });
+
+    it("should reject if the list iOS apps fails", async () => {
+      const expectedError = new Error("HTTP Error 404: Not Found");
+      const listAppsStub = listAppsApiRequestStub(AppPlatform.IOS).rejects(expectedError);
+
+      let err;
+      try {
+        await listFirebaseApps(PROJECT_ID, AppPlatform.IOS);
+      } catch (e) {
+        err = e;
+      }
+      expect(err.message).to.equal(
+        "Failed to list Firebase IOS apps. See firebase-debug.log for more info."
+      );
+      expect(err.original).to.equal(expectedError);
+      expect(listAppsStub).to.be.calledOnce;
+    });
+
+    it("should reject if the list Android apps fails", async () => {
+      const expectedError = new Error("HTTP Error 404: Not Found");
+      const listAppsStub = listAppsApiRequestStub(AppPlatform.ANDROID).rejects(expectedError);
+
+      let err;
+      try {
+        await listFirebaseApps(PROJECT_ID, AppPlatform.ANDROID);
+      } catch (e) {
+        err = e;
+      }
+      expect(err.message).to.equal(
+        "Failed to list Firebase ANDROID apps. See firebase-debug.log for more info."
+      );
+      expect(err.original).to.equal(expectedError);
+      expect(listAppsStub).to.be.calledOnce;
+    });
+
+    it("should reject if the list Web apps fails", async () => {
+      const expectedError = new Error("HTTP Error 404: Not Found");
+      const listAppsStub = listAppsApiRequestStub(AppPlatform.WEB).rejects(expectedError);
+
+      let err;
+      try {
+        await listFirebaseApps(PROJECT_ID, AppPlatform.WEB);
+      } catch (e) {
+        err = e;
+      }
+      expect(err.message).to.equal(
+        "Failed to list Firebase WEB apps. See firebase-debug.log for more info."
+      );
+      expect(err.original).to.equal(expectedError);
+      expect(listAppsStub).to.be.calledOnce;
+    });
+
+    function getListAppsResourceString(projectId: string, platform?: AppPlatform): string {
```

Comment:
This is a huge code-smell. Don't re-implement logic that exists in code in your tests. It just means that you have to maintain both implementations.

This is really making me reconsider these tests. Basically, you've created logic that only works with specific arguments. That's fine, but what happens when you call without those arguments is that the stub will return nothing (`undefined`) and break your test that way. I'm not sure that's _really_ the behavior you want. Technically, the code is doing the right thing, it returns what the stub returns.

Maybe a more useful setup might be this (I'll use an example from above - comments mostly for verbosity):

```typescript
    // Way above...:
    beforeEach(() => {
      apiRequestStub = sandbox.stub(api, "request").throws("Unexpected API request call");
    });

    it("should resolve with Android app list", async () => {
      // Generate the list that the stub should return when it's called:
      const expectedAppList = generateAppList(10, AppPlatform.ANDROID);
      // Now, set up the stub to return that list on the first call. Any other call will throw (from above).
      apiRequestStub.onFirstCall().resolves({ body: expectedAppList });

      // Pull this line out so it's very clear that this is the action we are taking/testing.
      const apps = await listFirebaseApps(PROJECT_ID, AppPlatform.ANDROID)

      expect(apps).to.deep.equal(expectedAppList);
      expect(listAppsStub).to.be.calledOnceWith(
        "GET",
        "/v1beta1/projects/${PROJECT_ID}/androidApps?pageSize=100",
        // `calledOnceWith` doesn't care about extra parameters (I'm fairly sure). In this case, it's
        // actually not important what the third parameter is, so it's not worth writing the assertion.
        // (There may be a need for one test that includes this third param, but not every one.)
      );
    });
```

Now that I've written this out, I actually like this a _lot_. Here's the benefits:
1. Nobody has to guess what the stub is supposed to be doing. It's defined right here that it will resolve the list on the first call.
1. If the stub _is_ called more than once, it will throw an error which is actually _useful_. The test will fail with "Unexpected API request call" in the error message (which can be tuned, maybe "Unexpected API request stub call") which makes it quick to debug.
1. The action that is taken is clear and well defined.
1. The expectation that the stub is called with the correct parameters is built into the test, not a helper function. This would remove the helper function (so all the logic stays in the test, in the best context) and actually improves the fact that _this test is asserting that the stub was called with the correct arguments_. NOT that the logic that exists in the stub generator helper function AND the test file are the same (which is a code-smell -  it's redundant).

I know that you have some other PRs in-flight that have the same logic. I strongly believe this pattern I've laid out above is much stronger in terms of readability and good practice. I would _like_ for you to make this improvement across all those PRs, but I know that can take some time and re-reviewing. Since this is going into a working branch for _you_, I wouldn't be opposed to updating the existing (and other) tests in another PR, but I certainly could see these tests blocking a merge into master.

If you have more thoughts or want to discuss this further, please reach out to me (`bkend@`). I'll be happy to talk through this more with you.

Reply:
Thanks Bryan! This is another great suggestion. I actually started without helper functions and what you saw here are my attempts refactoring duplicate code. I actually had some difficulties finding the best practice writing tests with `chai` and `sinon`, since we can have different ways writing valid test suites (and I'm not really familiar with the library before). 

Btw, I really like your suggestion, `callOnceWith` is really the key here, and I like the fact that it increases readability of my test a lot. 

I think I will left the tests as they are now, and send another PR changing all the written tests.

Reply:
Updated unit tests